python/ray/serve/_private/router.py

```diff
@@ -993,21 +989,37 @@ def __init__(
                     LongPollNamespace.RUNNING_REPLICAS,
                     deployment_id,
                 ): self._replica_scheduler.update_running_replicas,
+                (
+                    LongPollNamespace.AUTOSCALING_CONFIG,
+                    deployment_id,
+                ): self.update_autoscaling_config,
             },
             call_in_event_loop=event_loop,
         )
 
-        # Start the metrics pusher if autoscaling is enabled.
-        deployment_route = DeploymentRoute.FromString(
-            ray.get(controller_handle.get_deployment_info.remote(*deployment_id))
-        )
-        deployment_info = DeploymentInfo.from_proto(deployment_route.deployment_info)
         self.metrics_pusher = None
-        if deployment_info.deployment_config.autoscaling_config:
+        self.autoscaling_enabled = False
+        self.push_metrics_to_controller = controller_handle.record_handle_metrics.remote
+
+    def update_autoscaling_config(self, autoscaling_config):
+        self.autoscaling_config = autoscaling_config
+
+        # Start the metrics pusher if autoscaling is enabled.
+        if self.autoscaling_config:
             self.autoscaling_enabled = True
-            self.push_metrics_to_controller = (
-                controller_handle.record_handle_metrics.remote
-            )
+
+            # Optimization for autoscaling cold start time. If there are
+            # currently 0 replicas for the deployment, and there is at
+            # least one queued query on this router, then immediately
+            # push handle metric to the controller.
+            if (
+                len(self._replica_scheduler.curr_replicas) == 0
+                and self.num_queued_queries
+            ):
+                self.push_metrics_to_controller(
+                    {self.deployment_id: self.num_queued_queries}, time.time()
+                )
```

Comment:
This is a remote actor call, so we need to `ray.get`/`await` it, or it won't do anything.

Reply:
Actually, I think this is false. Ray remote calls are executed eagerly:

```
>>> import ray
>>> ray.init()
2024-01-17 09:39:17,275	INFO worker.py:1752 -- Started a local Ray instance. View the dashboard at http://127.0.0.1:8265 
RayContext(dashboard_url='127.0.0.1:8265', python_version='3.8.12', ray_version='3.0.0.dev0', ray_commit='{{RAY_COMMIT_SHA}}', protocol_version=None)
>>> @ray.remote
... def f(): print("hi")
... 
>>> f.remote()
ObjectRef(c8ef45ccd0112571ffffffffffffffffffffffff0100000001000000)
>>> (f pid=62709) hi
```

However, one remaining concern is that `async` actors don't guarantee task execution order ([docs](https://docs.ray.io/en/latest/ray-core/actors/task-orders.html#asynchronous-or-threaded-actor)). That means this call could get executed after a future call to `self.push_metrics_to_controller`, which means the controller would receive autoscaling information out of order. @zcin do you know what the impact of this would be?

Reply:
Nice, then I think this is okay. cc @edoakes in case there's anything I missed.